pgproto3/frontend.go

```diff
@@ -36,78 +38,152 @@ type Frontend struct {
 	rowDescription       RowDescription
 }
 
-func NewFrontend(r io.Reader, w io.Writer) (*Frontend, error) {
-	cr := chunkreader.NewChunkReader(r)
-	return &Frontend{cr: cr, w: w}, nil
+func NewFrontend(r io.Reader, w io.Writer, rawConn syscall.RawConn) (*Frontend, error) {
+	// By historical reasons Postgres currently has 8KB send buffer inside,
+	// so here we want to have at least the same size buffer.
+	// @see https://github.com/postgres/postgres/blob/249d64999615802752940e017ee5166e726bc7cd/src/backend/libpq/pqcomm.c#L134
+	// @see https://www.postgresql.org/message-id/0cdc5485-cb3c-5e16-4a46-e3b2f7a41322%40ya.ru
+	rb := rbuf.NewFixedSizeRingBuf(8192)
+	return &Frontend{rb: rb, w: w, rawConn: rawConn}, nil
 }
 
 func (b *Frontend) Send(msg FrontendMessage) error {
 	_, err := b.w.Write(msg.Encode(nil))
 	return err
 }
 
-func (b *Frontend) Receive() (BackendMessage, error) {
-	header, err := b.cr.Next(5)
-	if err != nil {
-		return nil, err
+// function to batch receive backend message
+// given array(msgs, msgBodies) must be already allocated
+// non-decoded BackendMessage will be assigned to msgs, and message body ([][]byte) will be assigned to msgBodies
+// n is the maximum row count to get data row
+func (b *Frontend) Receive(n int, msgs []BackendMessage, msgBodies [][]byte) (int, error) {
+	if n < 0 || n > len(msgs) || n > len(msgBodies) {
+		return 0, errors.Errorf("invalid array length, n: [%d], msg: [%d], msgBodies: [%d]", n, len(msgs), len(msgBodies))
 	}
 
-	msgType := header[0]
-	bodyLen := int(binary.BigEndian.Uint32(header[1:])) - 4
-
-	var msg BackendMessage
-	switch msgType {
-	case '1':
-		msg = &b.parseComplete
-	case '2':
-		msg = &b.bindComplete
-	case '3':
-		msg = &b.closeComplete
-	case 'A':
-		msg = &b.notificationResponse
-	case 'C':
-		msg = &b.commandComplete
-	case 'd':
-		msg = &b.copyData
-	case 'D':
-		msg = &b.dataRow
-	case 'E':
-		msg = &b.errorResponse
-	case 'G':
-		msg = &b.copyInResponse
-	case 'H':
-		msg = &b.copyOutResponse
-	case 'I':
-		msg = &b.emptyQueryResponse
-	case 'K':
-		msg = &b.backendKeyData
-	case 'n':
-		msg = &b.noData
-	case 'N':
-		msg = &b.noticeResponse
-	case 'R':
-		msg = &b.authentication
-	case 'S':
-		msg = &b.parameterStatus
-	case 't':
-		msg = &b.parameterDescription
-	case 'T':
-		msg = &b.rowDescription
-	case 'V':
-		msg = &b.functionCallResponse
-	case 'W':
-		msg = &b.copyBothResponse
-	case 'Z':
-		msg = &b.readyForQuery
-	default:
-		return nil, errors.Errorf("unknown message type: %c", msgType)
-	}
+	idx := 0 // the current index of messages(msgs)
+	rowCounts := 0
 
-	msgBody, err := b.cr.Next(bodyLen)
-	if err != nil {
-		return nil, err
-	}
+	header := make([]byte, 5) // current processing header
+	headerWp := 0             // the write position of header
+
+	var msgBody []byte // current processing message body
+	msgWp := 0         // the write position of message body
+
+	for {
+		// read all data from RawConn
+		if b.rb.Avail() <= 0 {
```

Comment:
shouldn't need to check this, when this code runs it should fill up the ring buffer it doesn't matter if there are available data to read or not. 

The loop should be to:
1. read everything you can until full or nothing left
2. decode every message until it cannot decode anymore as the message capacity is reached or no more messages to decode

so there should only be 2 loops here one after another, then return with number read, and it can be 0 (if connection didn't get any messages)